### ISSUE TITLE
[firebase auth] verify phone number not working on release application flutter

### ISSUE THREAD

**Issue by EngrTAHAR-Noureddine** [firebase/flutterfire#10139]
Hi Firebase Flutter developper,
i got this error when i implement package :  [firebase_auth 4.2.1](https://pub.dev/packages/firebase_auth).
this package work fine in debug mode but in release mode not working and i got this error :

```
E/flutter (29344): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: type 'Null' is not a subtype of type 'String'
E/flutter (29344): #0      yB.Rmc.<anonymous closure> (VDg:660)
E/flutter (29344): #1      _RootZone.PUa (Mtg:1586)
E/flutter (29344): #2      _BufferingStreamSubscription._Thb (Htg:339)
E/flutter (29344): #3      _BufferingStreamSubscription._add (Htg:271)
E/flutter (29344): #4      _Yb._add (Itg:123)
E/flutter (29344): #5      _bc._Wgb (Itg:253)
E/flutter (29344): #6      _Yb._Wgb (Itg:153)
E/flutter (29344): #7      _RootZone.PUa (Mtg:1586)
E/flutter (29344): #8      _BufferingStreamSubscription._Thb (Htg:339)
E/flutter (29344): #9      _Pb.Rhb (Htg:515)
E/flutter (29344): #10     _Sb.hhb (Htg:620)
E/flutter (29344): #11     _Sb.ehb.<anonymous closure> (Htg:591)
E/flutter (29344): #12     _oj (Etg:40)
E/flutter (29344): #13     _pj (Etg:49)
```

This is the code : 
```
Future<bool?> loginUser(String phone, BuildContext context) async{
    FirebaseAuth _auth = FirebaseAuth.instance;

    _auth.verifyPhoneNumber(
        phoneNumber: phone,
        timeout: const Duration(seconds: 60),
        verificationCompleted: (AuthCredential credential) async{
          Navigator.of(context).pop();

          UserCredential result = await _auth.signInWithCredential(credential);

          User? user = result.user;

          if(user != null){
            Navigator.push(context, MaterialPageRoute(
              builder: (context) => HomeScreen(user: user,)
            ));
          }else{
            print("Error");
          }

          //This callback would gets called when verification is done auto maticlly
        },
        verificationFailed: (FirebaseAuthException exception){
          print(exception);
        },
        codeSent: (String verificationId, int? forceResendingToken){
          showDialog(
            context: context,
            barrierDismissible: false,
            builder: (context) {
              return AlertDialog(
                title: const Text("Give the code?"),
                content: Column(
                  mainAxisSize: MainAxisSize.min,
                  children: <Widget>[
                    TextField(
                      controller: _codeController,
                    ),
                  ],
                ),
                actions: <Widget>[
                  TextButton(
                    child: const Text("Confirm", style: TextStyle(color: Colors.white, backgroundColor: Colors.blue),),
                    onPressed: () async{
                      final code = _codeController.text.trim();
                      AuthCredential credential = PhoneAuthProvider.credential(verificationId: verificationId, smsCode: code);

                      UserCredential result = await _auth.signInWithCredential(credential);

                      User? user = result.user;

                      if(user != null){
                        Navigator.push(context, MaterialPageRoute(
                            builder: (context) => HomeScreen(user: user,)
                        ));
                      }else{
                        print("Error");
                      }
                    },
                  )
                ],
              );
            }
          );
        },
        codeAutoRetrievalTimeout: (val){
          print(val);
        }
    );
  }
```

please help me ! this is urgent and search a lot and i can't find a solution.

**Comment by mariopepe** [comment]
I am getting this exact same issue…

**Comment by mariopepe** [comment]
There is an issue with phone number auth, is difficult to replicate or to reproduce, happens to Android in production when signing up a completely new user and AppAttest or Verification is on.

This is another instance: https://github.com/firebase/flutterfire/issues/10012

**Comment by darshankawar** [comment]
@EngrTAHAR-Noureddine 
Looking at the error you posted, `Unhandled Exception: type 'Null' is not a subtype of type 'String'`, probably indicates a null safety related issue. The default types (String, int, bool, double, etc) do not accept null as value to a variable, can you confirm that your code is using  Nullable types by appending a ? to the type like:

`String? myNullableString;`

If you think you are handling the null safety properly in your code, then provide use your `flutter doctor -v`. Also, does this error occur locally in your environment if running in `--release` mode ? If so, please provide steps to replicate as when does this error is triggered.

Also, Is this occuring on Android or iOS platform ?

**Comment by mariopepe** [comment]
@darshankawar you are wrong.

**About the method: (not just about this specific issue)**
Asking for for Flutter Doctor in this context is completely useless.
Also, you should understand that you are the maintainer for a package for which errors are often difficult to replicate as they have many outside services connected.
- is it in production?
- is the AppCheck / AppAttest enabled?
- is it Android or iOS?
- is the database correct?
- are the firestore rules right?
- moreover, providing a sample code is complex because authentication flow is deep inside the app structure, providing samples is not easy nor straight forward
- etc

Hence, asking for reproducible samples as you where the maintainer of some UI element package, such as an animated button, where you ask: "Hey show me the 3 lines in which you implement the button", is wrong.

**About the content:**
You said: "Looking at the error you posted, Unhandled Exception: type 'Null' is not a subtype of type 'String', probably indicates a null safety related issue"

**This is wrong, you ignore fix [10119](https://github.com/firebase/flutterfire/pull/10119)**, which resulted in a new version of `firebase_auth_platform_interface`(specifically version 6.11.5) which is likely creating the issue that the O.P. is reporting.

_(Here it was another mistake from the team, I honestly did not understand the reasoning behind it. You should have created a new non-breaking release for `firebase_auth`, that would have deployed the fix to users by doing a simple `flutter pub get` (thanks to caret operator), instead, you guys only released an update of `firebase_auth_platform_interface` and so now, to get the fixed version, the only way is to run `flutter pub upgrade`, ngl weird.)_

Anyway, this is the change-log...

<img width="509" alt="Screenshot 2022-12-19 at 13 26 46" src="https://user-images.githubusercontent.com/14803862/208425944-92802b76-8497-42ff-bf28-7b97613f5aba.png">

---
Message for @EngrTAHAR-Noureddine : to fix your issue you can do one of the following two things:
- add an explicit dependency in your pubspec.yaml file on: firebase_auth_platform_interface: ^6.11.5
- `run flutter pub upgrade` and be sure that this dependency is updated (honestly I would go for the method above).
---

Now, let's go back to the broader issue, @darshankawar.

[10119](https://github.com/firebase/flutterfire/pull/10119) was **reported AND fixed** by someone outside FlutterFire. So it means that your team was not aware of that.
It‘s fine, it can happen, but then when someone takes the time to report an issue, you guys should look into that waaaaaay more carefully, and assume that the client is right (since most of the times actually is).

Probably the bug 10119 was reported by someone in the past, but then you asked for Flutter Doctor and pointed the direction somewhere else, and it was not fixed. That is not nice, nor good.

**I honestly think that your way of monitoring the issues on this repo is slowing the ability of the team to find and fix issues.**

Let's look at another example, ([10012](https://github.com/firebase/flutterfire/issues/10012)) is another issue where the user was having issues with the phone number authentication (probably was this same issue), in which you tried the standard sample app, and the replied "I am unable to replicate this error using plugin's example"....

Now, any critique needs to be constructive in order not to be useful, so my constructive feedback is:
- you might start keeping counts of issues for the same package and the same type, so if 10 people open an issue for firebase_auth and specifically phone verification, **even if they do not provide the dang flutter doctor**, you still can tell the team "hey guys, I got 10 reports of issues of firebase_auth & phone authentication, I could not replicate, but can you investigate a bit? Likely there is a problem" (assume that if someone opens an issue is because they genuinely have a problem)
- get another pair of eyes (a collaborator) on the issues, we are humans, we can make mistakes, so being in 2 is better that being alone at doing complex tasks

The thing is that, from a service (firebase) I pay for good moneys, I would expect a different handling of the issues.
CC @Lyokone @russellwheatley

**Comment by Lyokone** [comment]
@mariopepe Thanks for your detailed message. I should have updated the message on the changelog for something more explicit.

I see that you have also commented on the PR. As you mention here, replicating a bug is hard since many variables are to be considered.

We often get issues that are due to improper configuration (Firebase is beginner friendly), hence why we try to have a process (where we ask for `flutter doctor` for instance). 

We'll review the triaging process to see if things can be improved.

**Comment by darshankawar** [comment]
Thanks for the feedback @mariopepe 

> **bout the method: (not just about this specific issue)** Asking for for Flutter Doctor in this context is completely useless. Also, you should understand that you are the maintainer for a package for which errors are often difficult to replicate as they have many outside services connected.
> 
> * is it in production?
> * is the AppCheck / AppAttest enabled?
> * is it Android or iOS?
> * is the database correct?
> * are the firestore rules right?
> * moreover, providing a sample code is complex because authentication flow is deep inside the app structure, providing samples is not easy nor straight forward
> * etc
> 
> Hence, asking for reproducible samples as you where the maintainer of some UI element package, such as an animated button, where you ask: "Hey show me the 3 lines in which you implement the button", is wrong.

As a first line of defense on triaging issues, we always look for all relevant details of the issue which helps to analyze the issue better and  then we are in good position to forward the issue to the core team so that they will have all necessary info to properly diagnose and possible fix the issue. There's a [new issue template](https://github.com/firebase/flutterfire/issues/new/choose) for a reason which contains all details that we would most likely need to make the issue actionable. If the issue is filed just with an error message, a code snippet that contains custom implementation with third party plugin code and then asking for help on the solution, then we are most likely going to ask for all details that are needed first before we can make a decision which is what I did in this case.



> You said: "Looking at the error you posted, Unhandled Exception: type 'Null' is not a subtype of type 'String', probably indicates a null safety related issue"
> 
> **This is wrong, you ignore fix [10119](https://github.com/firebase/flutterfire/pull/10119)**,

This is correct that the issue is most likely due to the fix you mentioned, but looking at the error posted about `null is not a subtype of string`, we first need to check / ask the OP if it is not due to a code implementation and they have properly used null safety, confirming this rules out the code implementation and then we can move forward to core issue.



> Let's look at another example, ([10012](https://github.com/firebase/flutterfire/issues/10012)) is another issue where the user was having issues with the phone number authentication (probably was this same issue), in which you tried the standard sample app, and the replied "I am unable to replicate this error using plugin's example"....

This issue again lacked most of the relevant details we need as I mentioned earlier and also considering the fact that the OP mentioned it wasn't replicable locally and only occured in production, that makes things a bit tough for us to properly analyze based on the details provided and hence we ask for all the details that we think could be helpful to narrow down the issue and that's what I did there and to tried to replicate. Please be aware of the fact that, using flutterfire plugins in a full fledged production app with so many modules and implementations present in it, trying to narrow down the issue to a particular plugin implementation developed and maintained by flutterfire team is critical and that's why we ask for details.


> * you might start keeping counts of issues for the same package and the same type, so if 10 people open an issue for firebase_auth and specifically phone verification, **even if they do not provide the dang flutter doctor**, you still can tell the team "hey guys, I got 10 reports of issues of firebase_auth & phone authentication, I could not replicate, but can you investigate a bit? Likely there is a problem" (assume that if someone opens an issue is because they genuinely have a problem)

We have been doing this already. We also take a look at number of comments received on an issue that is causing trouble to lot of devs and also the upvotes received on the issue (if any).  See [this](https://github.com/firebase/flutterfire/issues/10031#issuecomment-1337139150) issue comment for your reference.



> * get another pair of eyes (a collaborator) on the issues, we are humans, we can make mistakes, so being in 2 is better that being alone at doing complex tasks

Yes, if first line of defense or primary triage is unable to come to a conclusion, we do tag the concerned flutterfire team member and also bring to their attention so that they are aware of issues and the team also responds quickly.

**Comment by darshankawar** [comment]
Since a [PR](https://github.com/firebase/flutterfire/pull/10119) was already raised and has been merged fixing this issue, I have labeled this issue directly for tracking.

**Comment by EngrTAHAR-Noureddine** [comment]
@mariopepe  Thank you for your reply, i tried to fix this problem but still not working, 
i upgrade `firebase_auth_platform_interface`  and  i run `flutter pub upgrade` and get new error which is : `I/flutter (14255): unknown` ( try to register new user ( number phone exists )   )
the problem is in Android release  ( iOS works fine and in Android debug works fine too ).  
@darshankawar  i run `flutter doctor -v` it works fine and all is green check.
help me please and thank you to reply me .

**Comment by EngrTAHAR-Noureddine** [comment]
any updates

**Comment by Lyokone** [comment]
@EngrTAHAR-Noureddine I tried to reproduce the error in release mode on Android, but it's working fine on my end. Can you try to run `flutterfire update` to ensure that every FlutterFire plugin is properly updated on your end?

**Comment by EngrTAHAR-Noureddine** [comment]
@Lyokone thank you for your answer. 
I will try what you said.

**Comment by abddery** [comment]
I updated all firebase libraries to latest version then I add google play sha-1 and sha-256 to firebase console settings and it worked like a charm

**Comment by EngrTAHAR-Noureddine** [comment]
thank you for reply me @abddery, but i tried all solutions available and also yours but not working so i used a trick to solve this error **by creating a new project on firebase and associating my application to the new firebase project**. So i found that the old project of firebase not setting property. 
I thank @Lyokone @abddery @mariopepe @darshankawar for your replies.
